### PR TITLE
SALTO-2826: add retries to S3 operations

### PR DESCRIPTION
**Pull request by alonstern**
Added retries to ECONNREFUSED on S3 operations when flushing the state

---
_Release Notes_: 
_Core_:
- Added retries to ECONNREFUSED on S3 operations when flushing the state

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/core/src/local-workspace/s3_dir_store.ts [hunk 1]
```diff
@@ -27,6 +27,27 @@ const log = logger(module)
 
 const DEFAULT_CONCURRENCY_LIMIT = 100
 
+const RETRYABLE_MESSAGE = 'ECONNREFUSED'
+
+const NUMBER_OF_RETRIES = 5
+
+const runWithRetries = async <R>(
```
Comment (adamro) [inline packages/core/src/local-workspace/s3_dir_store.ts:34]: We don't have a fitting implementation already in the code-base? ie. lowerdash etc?

Comment (alonstern) [reply packages/core/src/local-workspace/s3_dir_store.ts:34]: I don't think so, but I might be missing something. It looks the the `retry` package in lowerdash expects the function to return something that is translated to false when it fails and not throws an error

**Comment by ori-moisis** [comment]
isn't this supported in the s3 client itself? can't this be solved by configuring the retry behavior that already exists?
https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html

File: packages/core/package.json [hunk 2]
```diff
@@ -33,6 +33,7 @@
   },
   "dependencies": {
     "@aws-sdk/client-s3": "^3.171.0",
+    "@aws-sdk/middleware-retry": "^3.190.0",
```
Comment (ori-moisis) [inline packages/core/package.json:36]: should this be on the same version as the rest of the aws-sdk stuff?
I think we've seen some issues in the past when the versions are inconsistent

**Review by ori-moisis — APPROVED** [review]
lgtm

**Pull request merged**